File: python/array_search/fractional_cascading_production.py
from bisect import bisect_left
from dataclasses import dataclass


@dataclass(frozen=True)
class Entry:
    value: int
    catalog_position: int
    next_position: int


@dataclass(frozen=True)
class CascadingQueryResult:
    positions: tuple[int, ...]
    matches: tuple[bool, ...]


class FractionalCascade:
    def __init__(self, catalogs):
        self.catalogs = tuple(tuple(catalog) for catalog in catalogs)
        self.layers = self._build_layers()

    def _build_layers(self):
        layers = [()] * len(self.catalogs)

        for index in range(len(self.catalogs) - 1, -1, -1):
            catalog = self.catalogs[index]
            next_layer = layers[index + 1] if index + 1 < len(layers) else ()
            sampled_next = next_layer[1::2]
            values = sorted(catalog + tuple(entry.value for entry in sampled_next))

            layers[index] = tuple(
                Entry(
                    value=value,
                    catalog_position=bisect_left(catalog, value),
                    next_position=self._entry_lower_bound(next_layer, value),
                )
                for value in values
            )

        return tuple(layers)

    @staticmethod
    def _entry_lower_bound(layer, target):
        left = 0
        right = len(layer)

        while left < right:
            mid = (left + right) // 2
            if layer[mid].value < target:
                left = mid + 1
            else:
                right = mid

        return left

    @staticmethod
    def _repair_position(layer, position, target):
        while position > 0 and layer[position - 1].value >= target:
            position -= 1

        while position < len(layer) and layer[position].value < target:
            position += 1

        return position

    def query(self, target):
        if not self.layers:
            return CascadingQueryResult((), ())

        positions = []
        matches = []
        position = self._entry_lower_bound(self.layers[0], target)

        for index, catalog in enumerate(self.catalogs):
            layer = self.layers[index]
            position = self._repair_position(layer, position, target)
            catalog_position = len(catalog) if position == len(layer) else layer[position].catalog_position

            positions.append(catalog_position)
            matches.append(catalog_position < len(catalog) and catalog[catalog_position] == target)

            if index + 1 < len(self.layers):
                position = len(self.layers[index + 1]) if position == len(layer) else layer[position].next_position

        return CascadingQueryResult(tuple(positions), tuple(matches))

File: java/array_search/FractionalCascadingProduction.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class FractionalCascadingProduction {
    public record QueryResult(List<Integer> positions, List<Boolean> matches) {
    }

    private record Entry(int value, int catalogPosition, int nextPosition) {
    }

    private final List<List<Integer>> catalogs;
    private final List<List<Entry>> layers;

    public FractionalCascadingProduction(List<List<Integer>> catalogs) {
        this.catalogs = catalogs.stream()
            .map(List::copyOf)
            .toList();
        this.layers = buildLayers();
    }

    public QueryResult query(int target) {
        List<Integer> positions = new ArrayList<>();
        List<Boolean> matches = new ArrayList<>();

        if (layers.isEmpty()) {
            return new QueryResult(positions, matches);
        }

        int position = entryLowerBound(layers.get(0), target);

        for (int index = 0; index < catalogs.size(); index++) {
            List<Entry> layer = layers.get(index);
            List<Integer> catalog = catalogs.get(index);

            position = repairPosition(layer, position, target);
            int catalogPosition = position == layer.size()
                ? catalog.size()
                : layer.get(position).catalogPosition();

            positions.add(catalogPosition);
            matches.add(catalogPosition < catalog.size() && catalog.get(catalogPosition) == target);

            if (index + 1 < layers.size()) {
                position = position == layer.size()
                    ? layers.get(index + 1).size()
                    : layer.get(position).nextPosition();
            }
        }

        return new QueryResult(positions, matches);
    }

    private List<List<Entry>> buildLayers() {
        List<List<Entry>> result = new ArrayList<>(Collections.nCopies(catalogs.size(), List.of()));

        for (int index = catalogs.size() - 1; index >= 0; index--) {
            List<Integer> catalog = catalogs.get(index);
            List<Entry> nextLayer = index + 1 < catalogs.size() ? result.get(index + 1) : List.of();
            List<Integer> values = new ArrayList<>(catalog);

            for (int nextIndex = 1; nextIndex < nextLayer.size(); nextIndex += 2) {
                values.add(nextLayer.get(nextIndex).value());
            }

            Collections.sort(values);

            List<Entry> layer = new ArrayList<>();
            for (int value : values) {
                layer.add(new Entry(
                    value,
                    lowerBound(catalog, value),
                    entryLowerBound(nextLayer, value)
                ));
            }

            result.set(index, List.copyOf(layer));
        }

        return List.copyOf(result);
    }

    private static int lowerBound(List<Integer> values, int target) {
        int low = 0;
        int high = values.size();

        while (low < high) {
            int mid = low + (high - low) / 2;

            if (values.get(mid) < target) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }

        return low;
    }

    private static int entryLowerBound(List<Entry> layer, int target) {
        int low = 0;
        int high = layer.size();

        while (low < high) {
            int mid = low + (high - low) / 2;

            if (layer.get(mid).value() < target) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }

        return low;
    }

    private static int repairPosition(List<Entry> layer, int position, int target) {
        while (position > 0 && layer.get(position - 1).value() >= target) {
            position--;
        }

        while (position < layer.size() && layer.get(position).value() < target) {
            position++;
        }

        return position;
    }
}
